engine/input.asm

```diff
@@ -4,82 +4,131 @@
 
 nop
 
+.struct InputState
+    index  db
+    start  db
+    select db
+    upbtn  db
+    dnbtn  db
+    lftbtn db
+    rhtbtn db
+.endst
+
 .struct Input
-    frame_counter dw
-    input1l_q instanceof Queue
-    input1h_q instanceof Queue
+    inputstate instanceof InputState
+    enabled db
 .endst
 
 .enum $00
     input instanceof Input
 .ende
 
-;Defining and init Queue here does not seem to be pushing/pop
-;Defining and init in Main seems to work for push, but it keeps pushing over memory
-;Queue_Define("INPUT1L_Q", $0090, #$0002, #$0004)
-;Queue_Define("INPUT1H_Q", $00A0, #$0002, #$0004)
-
 Input_Init:
-    pha
-    phy
-    phx
-
-    ; Allocate Queue + 4 bytes for the queue
-    lda #(_sizeof_Queue)
-    adc #$4
-    jsr Malloc_Bytes
-
-    ; X has the start address
-    ; Y has the end address
-    jsr Queue_Init
-
-    txy         ; Y is now the start address
-    lda 1, S    ; Grab the this pointer again
-    tax         ; X is now the proper offset
-    tya         ; Put the start address back into the accumulator
-    sta input.input1l_q, X ; Store the start address into the input struct
-
-    ; Allocate Queue + 4 bytes for the second queue
-    lda #(_sizeof_Queue)
-    adc #$4
-    jsr Malloc_Bytes
-
-    ; X has the start address
-    ; Y has the end address
-    jsr Queue_Init
-
-    txy         ; Y is now the start address
-    lda 1, S    ; Grab the this pointer again
-    tax         ; X is now the proper offset
-    tya         ; Put the start address back into the accumulator
-    sta input.input1h_q, X ; Store the start address into the input struct
-
-    lda #$0000
-    sta input.frame_counter, X
-
-    plx
-    ply
-    pla
+    A8
+    stz input.inputstate.upbtn, X
+    stz input.inputstate.dnbtn, X
+    stz input.inputstate.rhtbtn, X
+    stz input.inputstate.lftbtn, X
+    A16
     rts
 
 Input_Frame:
     rts
 
-; X is "this" pointer
 Input_VBlank:
     pha
-
-    inc input.frame_counter, X
-
-    lda JOY1L
-    call_ptr(Queue_Push, input.input1l_q)
-    call_ptr(Queue_Pop, input.input1l_q)
-
-    lda JOY1H
-    call_ptr(Queue_Push, input.input1h_q)
-    call_ptr(Queue_Pop, input.input1h_q)
-
+    jsr Input_UpButton
+    jsr Input_DnButton
+    jsr Input_LftButton
+    jsr Input_RhtButton
     pla
     rts
 
+Input_DnButton:
+    pha
+    @CheckDnButton:
+        lda JOY1L                          ; check whether the Dn button was pressed this frame...
+        cmp #DNBTN
+        bne @CheckDnButtonDone 
+        A8
+        lda #1
+        sta input.inputstate.dnbtn, X
+        bra @Done
+
+    @CheckDnButtonDone:
```

Comment:
Looking at your code, you could automatically populate your struct with a loop. The logic would probably follow:

1. Setup your input struct to have bytes ordered in the same way the bits are ordered in the low and high bytes of the standard controller data.
2. `X` register pointer should point to where `R` byte is in the struct.
3. Load the 16-bit word from the controller data and shift off `D0..D3` into your accumulator.
4. Setup a `Y` register to decrement down from `12` (since you shifted off the bits)
5. Loop over and test your LSB of the accumulator (`bit`). If true, then the current byte at X is 1, otherwise it is 0.
6. Increment your `X` register.
7. Right shift your accumulator.
8. Decrement your `Y` index register.
9. Loop until `Y` is `0`.

It might make it cleaner than manually checking each state.